src/Symfony/Bundle/FrameworkBundle/CacheWarmer/AbstractPhpFileCacheWarmer.php

```diff
@@ -0,0 +1,77 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\FrameworkBundle\CacheWarmer;
+
+use Psr\Cache\CacheItemPoolInterface;
+use Symfony\Component\Cache\Adapter\AdapterInterface;
+use Symfony\Component\Cache\Adapter\ArrayAdapter;
+use Symfony\Component\Cache\Adapter\PhpArrayAdapter;
+use Symfony\Component\Cache\Adapter\ProxyAdapter;
+use Symfony\Component\HttpKernel\CacheWarmer\CacheWarmerInterface;
+
+abstract class AbstractPhpFileCacheWarmer implements CacheWarmerInterface
+{
+    protected $phpArrayFile;
+    protected $fallbackPool;
+
+    /**
+     * @param string                 $phpArrayFile The PHP file where metadata are cached
+     * @param CacheItemPoolInterface $fallbackPool The pool where runtime-discovered metadata are cached
+     */
+    public function __construct($phpArrayFile, CacheItemPoolInterface $fallbackPool)
+    {
+        $this->phpArrayFile = $phpArrayFile;
+        if (!$fallbackPool instanceof AdapterInterface) {
+            $fallbackPool = new ProxyAdapter($fallbackPool);
+        }
+        $this->fallbackPool = $fallbackPool;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function warmUp($cacheDir)
+    {
+        $phpArrayAdapter = new PhpArrayAdapter($this->phpArrayFile, $this->fallbackPool);
+        $arrayAdapter = new ArrayAdapter();
+
+        spl_autoload_register(array($phpArrayAdapter, 'throwOnRequiredClass'));
+        try {
+            if (false === $this->doWarmUp($cacheDir, $phpArrayAdapter, $arrayAdapter)) {
+                return;
+            }
+        } finally {
+            spl_autoload_unregister(array($phpArrayAdapter, 'throwOnRequiredClass'));
+        }
+
+        // the ArrayAdapter stores the values serialized
+        // to avoid mutation of the data after it was written to the cache
+        // so here we un-serialize the values first
+        $values = array_map(function ($val) { return unserialize($val); }, array_filter($arrayAdapter->getValues()));
```

Comment:
@nicolas-grekas is it fine to apply `array_filter` here as well for the other cache warmers? Tests are fine although the `AnnotationsCacheWarmer` does not have any tests :cry: